File: pasterino-service/src/main/java/com/pyamsoft/pasterino/service/PasteServiceModule.java
/*
 * Copyright 2016 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.pasterino.service;

import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;
import com.pyamsoft.pasterino.base.PasterinoModule;
import rx.Scheduler;

public class PasteServiceModule {

  @NonNull private final PasteServiceInteractor interactor;
  @NonNull private final Scheduler obsScheduler;
  @NonNull private final Scheduler subScheduler;

  public PasteServiceModule(@NonNull PasterinoModule pasterinoModule) {
    interactor = new PasteServiceInteractor(pasterinoModule.providePreferences());
    obsScheduler = pasterinoModule.provideObsScheduler();
    subScheduler = pasterinoModule.provideSubScheduler();
  }

  @NonNull @CheckResult SinglePastePresenter getSinglePresenter() {
    return new SinglePastePresenter(interactor, obsScheduler, subScheduler);
  }
}

File: pasterino-service/src/main/java/com/pyamsoft/pasterino/service/PasteServiceModule.kt
/*
 *     Copyright (C) 2017 Peter Kenji Yamanaka
 *
 *     This program is free software; you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation; either version 2 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License along
 *     with this program; if not, write to the Free Software Foundation, Inc.,
 *     51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

package com.pyamsoft.pasterino.service

import android.support.annotation.CheckResult
import com.pyamsoft.pasterino.base.PasterinoModule
import io.reactivex.Scheduler

class PasteServiceModule(pasterinoModule: PasterinoModule) {

    private val interactor: PasteServiceInteractor
    private val pasteBus = pasterinoModule.providePasteBus()
    private val computationScheduler: Scheduler = pasterinoModule.provideComputationScheduler()
    private val ioScheduler: Scheduler = pasterinoModule.provideIoScheduler()
    private val mainScheduler: Scheduler = pasterinoModule.provideMainScheduler()

    init {
        interactor = PasteServiceInteractorImpl(pasterinoModule.providePreferences())
    }

    @CheckResult
    fun getSinglePresenter(): SinglePastePresenter =
            SinglePastePresenter(interactor, computationScheduler, ioScheduler, mainScheduler)

    @CheckResult
    fun getPasteServicePresenter(): PasteServicePresenter =
            PasteServicePresenter(pasteBus, computationScheduler, ioScheduler, mainScheduler)

    @CheckResult
    fun getPasteServicePublisher(): PasteServicePublisher =
            PasteServicePublisher(pasteBus)
}
